63515 added output of a JSON-formatted metrics representing the entire run

MatthiasLeeIBM opened the pull request:
Bugzid:  63515

Changes:
- added another histogram for collecting stats across the entire run
- added JSON-formatted output of collected statistics

Generated Output:
```json
{ "recordedMetrics" : {
"put": {"n":319546,
"standard_deviation":0.9843117238528302,
"median":2,
"harmonic_mean":2.2260478373026475,
"geometric_mean":2.2784470685780045,
"arithmetic_mean":2.3607399247682648,
"min":2,
"max":93,
"p50":2,
"p75":3,
"p90":3,
"p95":3,
"p99":5,
"p999":15,
"variance":0.9688695697141303,
"kurtosis":1163.4918872449086,
"skewness":21.457948218223454
},
"put10": {"n":282077,
"standard_deviation":3.6205383023363447,
"median":17,
"harmonic_mean":17.28713647538748,
"geometric_mean":17.4556781472093,
"arithmetic_mean":17.693165341378418,
"min":15,
"max":242,
"p50":17,
"p75":18,
"p90":20,
"p95":24,
"p99":32,
"p999":50,
"variance":13.10829759868454,
"kurtosis":134.8276758531883,
"skewness":6.980396917158458
}
}
}
```


Questions:
- Is _slide_ the right histogram for this? looks like _slide_ only represents the last window-size worth of stats, might be better to do a _uniform_ histogram.
- I'd be interested in pointers on style and best practices.

cc: @chewbranca @djschlegel2

chewbranca commented:
@MatthiasLeeIBM yeah you definitely want to use a `uniform` histogram to get stats from the full run.

Thinking about it more, I'm not sure saving this output in JSON is the right approach. Really it seems like the format we want to have direct access to is tab separated output (or even csv) so that we've got one key/val pair per line and can easily grep through it. Using JSON from the command line can be annoying.

Also having JSON data written from the `csv` stats writer is a bit awkward, but I'm not sure there's a better place. If you do really want to output in JSON, let's make it easy to switch between the two in a config setting so we can get a command line friendlier version as well.

Also, I realized that we've got access to a JSON encoder without any additional deps in the `mochijson` module. Let's switch to that rather than generating JSON strings [1].

[1] https://github.com/mochi/mochiweb/blob/master/src/mochijson.erl#L25-L26

MatthiasLeeIBM commented:
@chewbranca, I've changed it to a 16384bin uniform histogram.

Re JSON format: I think it's key to have a flexible format that we can insert directly into the database. A while back we standardized test results files destined for Nimbus to just JSON, that way it's easy to jam in any kind of settings/configurations/sysinfo without needing to change the format, ultimately I was hoping to also jsonify the whole set of configurations that were used at runtime. I'll figure out how to add a switch to get it to spit out the same info as a CSV file. 

I looked at mochijson but we [tore it out during the refactoring](https://github.com/cloudant/basho_bench/commit/150aef96ceadfae7c0d4ebbe5e2cbc2a06d0bfef) do you want me to add that dependency back in?

EDIT: after some discussion it was decided that jiffy is a better choice

MatthiasLeeIBM commented:
@chewbranca, when you get the chance, can you take another peek?

New Changes:
- added a CSV output, which can be switched on/off via the added config parameter: `run_statistics_output_format`
- fixed a bug, where the "n" value was just the size of the histogram and not the number of ops
- switched to using jiffy for JSON generation

chewbranca commented on src/basho_bench_stats.erl:
```diff
@@ -111,6 +113,8 @@ init([]) ->
     %% successful operations
     [begin
          folsom_metrics:new_histogram({latencies, Op}, slide, basho_bench_config:get(report_interval)),
+         folsom_metrics:new_histogram({overall_latencies, Op}, uniform, 16384),
```
Why 16384?

MatthiasLeeIBM replied:
Originally I started with 1024, but I figured that if we collect p999, we should have some more datapoints than just ~1 that would fall around that percentile, so I figured approx 10x  and then round up to the next power of two.

Its a bit arbitrary, but that was my thought process

chewbranca commented on src/basho_bench_stats.erl:
```diff
@@ -67,6 +67,8 @@ op_complete(Op, {ok, Units}, ElapsedUs) ->
         false ->
             folsom_metrics:notify_existing_metric({latencies, Op}, ElapsedUs, histogram),
+            folsom_metrics:notify_existing_metric({overall_latencies, Op}, ElapsedUs, histogram),
+            folsom_metrics:notify_existing_metric({overall_units, Op}, {inc, Units}, counter),
```
What does `{overall_units, Op}` provide us? I'm not sure we need this, seems like we could get this same data by summing the `n` column values from the other data sets. Or is there something else this provides?

MatthiasLeeIBM replied:
So yea, we could sum over the `n` from the other data set, but the other data set keeps getting cleared after every window, meaning we'd have to keep track of it somewhere else. Seemed cleaner to just run a separate counter

chewbranca commented on src/basho_bench_stats.erl:
```diff
@@ -270,6 +279,15 @@ process_stats(Now, #state{stats_writer=Module}=State) ->
+        Stats1 = folsom_metrics:get_histogram_statistics({overall_latencies, Op}),
+        Units = folsom_metrics:get_metric_value({overall_units, Op}),
+        Module:report_global_stats(Op, Stats1, Units)
+        end, State#state.ops).
```
This last line should be at the same indentation level as the `lists:foreach(...)` call.

chewbranca replied:
Also, I think you can switch `Stats1` to just `Stats` as I don't see it used elsewhere in this function.

MatthiasLeeIBM replied:
oops, yea that was left from the previous iteration. fixed

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+    %% Open metrics file
+    TestDir = basho_bench:get_test_dir(),
+    {ok, GlobalMetricsFile} = file:open(
+        filename:join([TestDir, "runStatistics.json"]),
```
The rest of the result files are using underscore separators rather than camel case, let's stick with that pattern.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -88,6 +97,36 @@ report_error({_SummaryFile, ErrorsFile},
+            _ ->
+                [{K, proplists:get_value(K, Stats)} | Acc ]
+        end
+        end, [], proplists:get_keys(Stats)),
```
This should be indented four spaces to the left.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+    ),
+
+    %% final assembly of JSON and writing to file
+    file:write(GlobalMetricsFile, io_lib:format("~s\n", [jiffy:encode({[{recordedMetrics,{RunMetrics}}]}, [pretty])])),
```
This line probably needs to be divided up as it looks to be longer than 80 chars and is scrunched together. Also, is the `io_lib:format` call necessary here?

MatthiasLeeIBM replied:
I've fixed the formatting, I seemed to have to use io_lib, don't remember what the error was. I'll look into it again

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+    %% Open metrics file
+    TestDir = basho_bench:get_test_dir(),
+    {ok, GlobalMetricsFile} = file:open(
+        filename:join([TestDir, "runStatistics.csv"]),
```
Same comment about camel case here.

MatthiasLeeIBM replied:
yep fixed in both

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+    Headers = lists:foldl(fun(K, Acc) ->
+        {MetricsObj} = proplists:get_value(K, RunMetrics),
+        proplists:get_keys(MetricsObj) ++ Acc
+        end, [], proplists:get_keys(RunMetrics)),
```
Four space left indentation.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+        CsvLine = lists:foldl(fun(K, Acc) ->
+            StrVal = io_lib:format("~w",[proplists:get_value(K, MetricsObj, "")]),
+            [StrVal|Acc]
+            end,[],UniqueHeaders),
```
Should be four spaces to the left.

chewbranca replied:
Also need spaces between the params.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,59 @@ measurement_csv_file({Label, _Op}) ->
+
+        file:write(GlobalMetricsFile, 
+            io_lib:format("~s\n", [string:join([atom_to_list(Op)|CsvLine],", ")]))
+        end, proplists:get_keys(RunMetrics)),
```
Same with this line.

MatthiasLeeIBM replied:
fixed.

MatthiasLeeIBM commented:
@chewbranca:
I did some benchmarking to see what the effects are of adding another histogram and counter.
The results are as follows

**Values processed in 60s:**

|  | 1 counter | 2 counters |
| --- | --- | --- |
| 1 hist | ~20m | ~18m |
| 2 hist | ~15m | ~14.5m |

Sidenote: these tests were run on my laptop.

Since that is still ~240k Ops/sec even on my laptop, I think this is probably still plenty faster enough for us to be running even very large tests.

We could test splitting out one histogram per worker thread, but I havn't gotten around to that.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -88,6 +97,36 @@ report_error({_SummaryFile, ErrorsFile},
+                 {p95, proplists:get_value(95, P)},
+                 {p99, proplists:get_value(99, P)},
+                 {p999, proplists:get_value(999, P)}
+                 ] ++ Acc;
```
This should be four space indentation and the first tuple should be indented with the rest. For example:

```
                [
                    {p50, proplists:get_value(50, P)},
                    {p75, proplists:get_value(75, P)},
                    {p90, proplists:get_value(90, P)},
                    {p95, proplists:get_value(95, P)},
                    {p99, proplists:get_value(99, P)},
                    {p999, proplists:get_value(999, P)}
                ] ++ Acc;
```

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -88,6 +97,36 @@ report_error({_SummaryFile, ErrorsFile},
+
+            %% All other metrics get turned into {<key>, <value>}
+            _ ->
+                [{K, proplists:get_value(K, Stats)} | Acc ]
```
Extra space after `Acc`.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,53 @@ measurement_csv_file({Label, _Op}) ->
+    UniqueHeaders = lists:usort(Headers),
+    StringyfiedHeader =  io_lib:format("~s\n", 
+        [string:join( 
+           lists:map(fun atom_to_list/1, [op|UniqueHeaders]), ", ") ]),
```
I would split this function out into separate constructs as it's a bit confusing in its current form. Maybe something like:

```
StrHeaders = string:join(lists:map(fun atom_to_list/1, [op|UniqueHeaders]), ", "),
StringyfiedHeader =  io_lib:format("~s\n", [StrHeaders]),
```

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,53 @@ measurement_csv_file({Label, _Op}) ->
+        proplists:get_keys(MetricsObj) ++ Acc
+    end, [], proplists:get_keys(RunMetrics)),
+
+    %% Remove all duplicates and sort list of available headers, write to file
```
Outdated comment, this still mentions writing to a file.

MatthiasLeeIBM replied:
fixed

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,53 @@ measurement_csv_file({Label, _Op}) ->
+            [StrVal|Acc]
+        end, [], UniqueHeaders),
+
+        [io_lib:format("~s\n", [string:join([atom_to_list(Op)|CsvLine],", ")])|Lines]
```
Inconsistent whitespace in this line.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -133,6 +172,53 @@ measurement_csv_file({Label, _Op}) ->
+    
+    %% Write lines to file
+    lists:foreach(fun(Line) ->
+        file:write(GlobalMetricsFile, Line)
```
I would check the return value here like so: `ok = file:write(GlobalMetricsFile, Line)`.

MatthiasLeeIBM replied:
fixed.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -204,7 +204,8 @@ stringify_stats(_RunStatsType=csv, RunMetrics) ->
 
-        [io_lib:format("~s\n", [string:join([atom_to_list(Op)|CsvLine],", ")])|Lines]
+        [io_lib:format("~s\n", 
+            [string:join( [atom_to_list(Op) | CsvLine], ", ")]) | Lines]
```
This line is very confusing. It takes a fair bit of parsing to understand that it's appending into a list as opposed to a list comprehension or something else. I would do something along the lines of one of the following:

```
OpStr = string:join( [atom_to_list(Op) | CsvLine], ", "),
[io_lib:format("~s\n", [OpStr]) | Lines]
```

or

```
[
    io_lib:format("~s\n", [
        string:join( [atom_to_list(Op) | CsvLine], ", ")])
    ])
    | Lines
]
```

chewbranca replied:
Or a combination where you have `OpLine = io_lib:format("~s\n", [string:join( [atom_to_list(Op) | CsvLine], ", ")]),` and then `[OpLine | Lines]`.

MatthiasLeeIBM replied:
I went with the first variant, i like it more compact

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -204,8 +204,9 @@ stringify_stats(_RunStatsType=csv, RunMetrics) ->
 
-        [io_lib:format("~s\n", 
-            [string:join( [atom_to_list(Op) | CsvLine], ", ")]) | Lines]
+        OpStr = string:join( [atom_to_list(Op) | CsvLine], ", "),
```
Extra space after `string:join(`.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -186,16 +193,15 @@ dump_run_statistics(RunStatsType) ->
+    OrderedFields = [mean, geometric_mean, harmonic_mean,
+        variance, standard_deviation, skewness, kurtosis,
+        median, p75, p90, p95, p99, p999,
+        min, max, n, basho_errors],
```
Let's move `n` to the front of this list so that the rows start with `op, n, ...`. This will more closely lineup with the existing output format that puts `n` ahead of the rest of the stats.

MatthiasLeeIBM replied:
ok, done.

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -88,6 +97,46 @@ report_error({_SummaryFile, ErrorsFile},
+    end, [], proplists:get_keys(Stats)),
+
+    %% insert correct units count
+    JsonElements1 = lists:keyreplace(n, 1, JsonElements0, {'_n', Units}),
```
Is this a typo here with `_n`?

MatthiasLeeIBM replied:
not a typo, it's been something that we adopted at some point, a long time ago. 
It's been bugging me, but no more, i've changed it internally on nimbus. so I will be changing it to 'n'

chewbranca commented on src/basho_bench_stats_writer_csv.erl:
```diff
@@ -67,6 +71,11 @@ new(Ops, Measurements) ->
     ?INFO("module=~s event=stop stats_sink=csv\n", [?MODULE]),
+
+    %% write out run-wide statistics before exiting
+    RunStatsType = basho_bench_config:get(run_statistics_output_format, json),
```
Let's switch the default to csv as the rest of our stats default to outputting as csv.

MatthiasLeeIBM replied:
done

chewbranca commented:
A few minor nits but overall this looks good. +1 to squerge with the last few items addressed.
